app_vue/src/stores/route_store.js

```diff
@@ -1,79 +1,127 @@
+        this.selectedRouteList.splice(index, 1);
       }
-      // reset
-      this.enableOptimizeRoute = true;
-      this.isRouteOptimized = false;
     },
     clearSensorRoute() {
-      this.sensorRouteList = [];
+      this.selectedRouteList = [];
       // reset
-      this.enableOptimizeRoute = true;
       this.isRouteOptimized = false;
     },
-    setEnableOptimizedRoute(value) {
-      this.enableOptimizeRoute = value;
-    },
     setIsRouteOptimized(value) {
       this.isRouteOptimized = value;
     },
-    // reorders intermediate points in our sensorRouteList according to a list of given indexes (routeOrder)
+    isAlreadyInRoute(sensor) {
+      if (this.getSelectedRouteList.length > 0) {
+        return !!this.getSelectedRouteList.find(bin => bin.id === sensor.id);
+      }
+      return false
+    },
+    // reorders intermediate points in our selectedRouteList according to a list of given indexes (routeOrder)
     updateWithOptimizedRoute(routeOrder) {
       // assuming routeOrder looks like [originLocationSensor, midLocationSensor1...midLocationSensor4, destinationLocaitonSensor]
       // variable to hold our intermediate waypoints for update (no start/end point included)
-      const intermediates = [...this.sensorRouteList];
-      intermediates.shift(); // remove start point
-      intermediates.pop(); // remove end point
+      const intermediates = [...this.selectedRouteList];
 
       // refer to google route order and match it to our defined route
       // copy result to temp
       let temp = [];
-      const arrLength = intermediates.length;
-      for (let i = 0; i < arrLength; i++) {
-        temp[routeOrder[i]] = intermediates[i];
+      for (let i = 0; i < routeOrder.length; i++) {
+        temp[i] = intermediates[routeOrder[i]];
       }
 
       // replace intermediate waypoint array with temp
-      for (let j = 0; j < arrLength; j++) {
+      for (let j = 0; j < routeOrder.length; j++) {
         intermediates[j] = temp[j];
       }
 
       // update saved routelist with new route order
-      this.sensorRouteList.splice(1, intermediates.length, ...intermediates);
+      this.selectedRouteList = [...intermediates];
+    },
+    async googOptimizeRoute() {
+      const googResponse = await getOptimizedRoute(this.getSelectedRouteList, this.startPointAddress, this.endPointAddress);
+      if (googResponse && googResponse.routes && googResponse.routes[0]) {
+        const routeOrder = googResponse.routes[0].optimizedIntermediateWaypointIndex; // [0,3,4]
+        this.updateWithOptimizedRoute(routeOrder); // update current route
+        this.setIsRouteOptimized(true); // set flag is optimized to true
+
+        if (googResponse.routes[0].duration) {
+          this.setRouteDuration(googResponse.routes[0]?.duration);
+        }
+
+        if (googResponse.routes[0].distanceMeters) {
+          this.setRouteDistance(googResponse.routes[0]?.distanceMeters);
+        }
+        
+      }
```

Comment:
google api call is now moved to route store to be called from any component as needed